File: src/Asm.MooBank.Web.App/src/models/Instrument.ts
import { VirtualInstrument } from "./VirtualAccount";

export const Controllers = ["Manual", "Virtual", "Import"] as Controller[];
export type Controller = "Manual" | "Virtual" | "Import";

export type InstrumentId = string;

export interface Instrument {
    id: InstrumentId;
    name: string;
    controller: Controller;
    description?: string;
    currentBalance: number;
    currentBalanceLocalCurrency: number;
    currency: string;
    instrumentType?: string;
    virtualInstruments: VirtualInstrument[];
    remainingBalance?: number;
    remainingBalanceLocalCurrency?: number;
}

export interface TopLevelAccount {
    shareWithFamily: boolean;
    groupId: string;
}

File: src/Asm.MooBank/Models/Instrument.cs
namespace Asm.MooBank.Models;

public abstract record Instrument
{
    public required Guid Id { get; set; }

    public required string Name { get; set; }

    public string? Description { get; set; }

    public DateTimeOffset BalanceDate { get; set; } = DateTimeOffset.Now;

    public required decimal CurrentBalance { get; set; }

    public required decimal? CurrentBalanceLocalCurrency { get; set; }

    public required string Currency { get; set; }

    public required Controller Controller { get; init; }

    public DateOnly? ClosedDate { get; set; }

    public string? InstrumentType { get; set; }

    public Guid? GroupId { get; set; }

    public virtual ICollection<VirtualInstrument> VirtualInstruments { get; init; } = [];

    /// <summary>
    /// Gets the remaining balance once virtual instruments are accounted for.
    /// </summary>
    public decimal? RemainingBalance { get; init; }

    /// <summary>
    /// Gets the remaining balance once virtual instruments are accounted for.
    /// </summary>
    public decimal? RemainingBalanceLocalCurrency { get; init; }
}
